fix(ngOptions): add ngEmptyValue attribute to ngOptions

Thesoro opened the pull request:
Reference: angular#13334

This would allow an additional attribute on ng-options to provide a default value for options that lack values.

Thesoro committed:
feat(ngOptions): add ngEmptyValue attribute to ngOptions

allow an optional ng-empty-value attribute on select elements using ng-options. any option whose value is an empty string will be replaced with the given empty-value.

Closes: #13334

Narretz commented:
Thanks for the PR! Before we can merge we a few things:
[ ] Docs for the new feature
[ ] A test for the new feature
[ ] Squashing all commits

Can you work on improving this? If you need help with setting up a dev environment, check https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md or ask here

Thesoro commented:
Will do! Thanks for the instructions, I'll add docs and a test, then squash all the commits together.

I'm a little confused about commit messages - I thought I'd included the issue. Should I use "closes" rather than "references"?

Also, the original issue was opened as a bug, has it been reassessed as a feature? (No disagreement if it is! I'm just a newbie to the project and I want to make sure I understand)

Narretz commented:
"Closes" in the commit message closes the issue when your commit is merged into master. Which is what we want. :) I say it's a feature, because it's not a bug, it's just something that's missing. Or did this ever work before in some way?

Thesoro commented:
That all makes sense to me. I don't believe this has worked before, but I just searched for the "PRs Plz" tag and grabbed one I could make sense of.

Narretz commented:
Cool, it's always good to see new contributors. These formal issues (commit message, multiple commits) can be handled when code & docs are ready, they are not blocking.

Thesoro commented:
Thanks for your help. When correcting a previously made PR, should I add a comment to that effect?
If so, here is that comment.
If not, uh, never mind.

Narretz commented on test/ng/directive/ngOptionsSpec.js:
```diff
@@ -2143,6 +2143,25 @@ describe('ngOptions', function() {
+      setSelectValue(element, 3);
+      setSelectValue(element, 0);
+
+      //make sure the empty value is being evaluated as javascript
```
Angular expressions are only a subset of Javascript.

Narretz commented on test/ng/directive/ngOptionsSpec.js:
```diff
@@ -2143,6 +2143,25 @@ describe('ngOptions', function() {
     });
 
+    it('should be possible to specify a value for empty options', function() {
+      var html = '<select ng-model="someModel" ng-options="c for c in choices" ng-empty-value="2+2">' +
```
I think the attribute should reference a scope property. This is more in line with what people will do.

Narretz commented on src/ng/directive/ngOptions.js:
```diff
@@ -156,6 +159,9 @@ var ngOptionsMinErr = minErr('ngOptions');
  *      even when the options are recreated (e.g. reloaded from the server).
  *
+ * @param {expression=} ngEmptyValue Any empty options, whether hard-coded or
+ * created with ngOptions, will be given the value that this expression evaluates to.
```
The empty option is still `''` if I am not mistaken. The ngEmptyValue is passed to $setViewValue and then ends up as the model.

Narretz commented:
Yes, you should write a commit. Because subscribers don't get notifications for code pushes.

I wonder if the feature is "good enough", because it only works if the option is explicitly selected. I would expect it to work also on load, before any interaction has happened.

Thesoro commented:
Okay, I believe I addressed your comments. Thanks for your time in giving those!
I don't really have anything for the pageload problem. I agree that it'd be better if it worked before selection, but that looks like a much trickier fix (at least for someone like me who's still getting his bearings in the Angular source code).

Narretz commented:
Good stuff. I have to think about setting the empty option on load. It's probably not necessary to do that.
For the use case described in the original issue, it's not problem, because the model would probably be undefined until it is selected again in the options. The only case I could imagine where this might be useful is when you set your model to a value that is not in your options. In that case, we reset the model to the empty value. So that case is covered.

Narretz commented on test/ng/directive/ngOptionsSpec.js:
```diff
@@ -2143,6 +2143,26 @@ describe('ngOptions', function() {
     });
 
+    it('should be possible to specify a value for empty options', function() {
+      var html = '<select ng-model="someModel" ng-options="c for c in choices" ng-empty-value="someVar">' +
```
You should use the createSingleSelect fn here, to make it consistent with other tests

Thesoro replied:
If I use createSingleSelect, how do I give it an ng-empty-value attribute? I was reluctant to change such a heavily used function for just the one test.

Narretz replied:
True, didn't think about that. You should be able to use createSelect:
```
    createSelect({
      'ng-model':'selected',
     'ng-empty-value': 'someVar',
      'ng-options':'value.name for value in values'
    }, true);
```

Narretz commented on test/ng/directive/ngOptionsSpec.js:
```diff
@@ -2143,6 +2143,26 @@ describe('ngOptions', function() {
+      //make sure the empty value is being evaluated as an Angular expression
+      expect(scope.someModel).toBe("foo");
+
+      dealoc(element);
```
No need to dealoc

Narretz commented on test/ng/directive/ngOptionsSpec.js:
```diff
@@ -2143,6 +2143,26 @@ describe('ngOptions', function() {
+      setSelectValue(element, 0);
+
+      //make sure the empty value is being evaluated as an Angular expression
+      expect(scope.someModel).toBe("foo");
```
We should also check if a change to someVar is picked up by the options directive. So change someVar, change the model, then change back to empty option and check

Narretz commented on src/ng/directive/ngOptions.js:
```diff
@@ -489,6 +495,9 @@ var ngOptionsDirective = ['$compile', '$parse', function($compile, $parse) {
         selectCtrl.readValue = function readNgOptionsValue() {
 
+          // Allow a default value to be substituted for any options with empty values
+          if (selectElement.val() === '') return $parse(attr.ngEmptyValue)(scope);
```
The expression $parse should be moved out the readValue fn. It only has to be done once.
So above the function ,do something like `var emptyValueParseFn = 'ngEmptyValue' in attr && attr.ngEmptyValue.length > 0 && $parse(attr.ngEmptyValue);`

and inside the read, do 
`if (selectElement.val() === '' && emptyValueParseFn) return emptyValueParseFn(scope);`

This also ensures we don't use the emptyValueParseFn when the attribute is not defined or if it is empty.
It's a bit odd that this isn't caught by the tests. We should enhance one of the exisiting tests to make sure that the model value is `null` when no ng-empty-value attribute is found.

Thesoro replied:
I'm a little lost on which test to update for this - it appears that this test...
`    it('should update model to null on change', function() {
      createSingleSelect(true);

      scope.$apply(function() {
        scope.values = [{name: 'A'}, {name: 'B'}];
        scope.selected = scope.values[0];
      });

      element.val('');
      browserTrigger(element, 'change');
      expect(scope.selected).toEqual(null);
    });`
...already checks that the model value is null on empty values in a no ng-empty-value situation? It does so implicitly, though.

Narretz commented:
Sorry, I keep finding things. ;)

Thesoro committed:
changes in response to notes in PR thread -- to be squashed later

Thesoro commented:
Updated for most of your notes, with the exception of the one I replied about in your line notes. Thanks for your time in looking at these!

Thesoro committed:
split line for linter

cmplank commented:
So... is this PR dead? I was trying to use this feature until I realized it was never actually merged. :disappointed:

gkalpak commented:
TBH, now that `select` can be also used with `ngRepeat`/`ngValue`, I see little value in adding this (which is only useful in specific usecases). 

[1]: http://plnkr.co/edit/sp5NS6?p=preview

Narretz commented:
I agree that this is not needed in core. Although I don't see how ngValue + select changes the assumptions here - both regular and ngOptions select will set the model to `null` if the empty option is selected.

In https://github.com/angular/angular.js/issues/1780, there are 2 workarounds for this problem, and I've added a 3rd one that implements emptyValue, but doesn't rely on private API.
